spark/v3.4/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestMerge.java

```diff
@@ -2728,6 +2736,105 @@ public void testMergeToWapBranchWithTableBranchIdentifier() {
                         branch)));
   }
 
+  @Test
+  public void testMergeFilterWithSystemFunctionsOnPartitionedTable() {
+    Assumptions.assumeThat(catalogName).isNotEqualTo("spark_catalog");
+
+    Map<String, String> tableProperties = Maps.newHashMap();
+    tableProperties.put(DEFAULT_FILE_FORMAT, fileFormat);
+    tableProperties.put(WRITE_DISTRIBUTION_MODE, distributionMode);
+    tableProperties.put(SPARK_WRITE_PARTITIONED_FANOUT_ENABLED, String.valueOf(fanoutEnabled));
+    tableProperties.putAll(extraTableProperties());
+    createPartitionedTable(spark, tableName, "years(ts)", tableProperties);
+
+    createOrReplaceView(
+        "source",
+        "id BIGINT, ts TIMESTAMP, data STRING",
+        "{ \"id\": 0, \"ts\": \"2023-08-01T09:20:44.294658+00:00\", \"data\": \"new-data-0\"}\n"
+            + "{ \"id\": 5, \"ts\": \"2023-08-02T09:23:44.294658+00:00\", \"data\": \"new-data-5\"}\n");
+
+    String date = "2017-11-22";
+    String resultSql =
+        sqlFormat("SELECT id, ts, IF(id = 0, 'new-data-0', data) FROM %s ORDER BY id", tableName);
+    List<Object[]> expected = sql(resultSql);
+
+    withSQLConf(
+        ImmutableMap.of(SparkSQLProperties.SYSTEM_FUNC_PUSH_DOWN_ENABLED, "true"),
+        () -> {
+          String mergeSql =
+              sqlFormat(
+                  "MERGE INTO %s AS t USING source AS s "
+                      + "ON t.id = s.id AND %s.system.years(t.ts) = %s.system.years(date('%s')) "
+                      + "WHEN MATCHED "
+                      + "  THEN UPDATE SET t.data = s.data "
+                      + "WHEN NOT MATCHED AND s.id != 5 "
+                      + "  THEN INSERT *",
+                  tableName, catalogName, catalogName, date);
+          Dataset<Row> df = spark.sql(mergeSql);
+          // Should not contain applyfunction since filter function could be evaluated completely on
+          // Iceberg side and will not return post scan filters
+          Assertions.assertThat(df.queryExecution().sparkPlan().toString())
+              .doesNotContain("Filter (applyfunctionexpression(Wrapper(iceberg.years(timestamp))");
```

Comment:
Using `doesNotContain` is a very brittle assertion because it could be true if Spark changes the format of its string representation. Can you instead find a way to test more directly?

Reply:
Row-level commands UTs updated as well.